pkg/codegen/internal/test/testdata/simple-methods-schema/nodejs/foo.ts

```diff
@@ -0,0 +1,117 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+import * as pulumi from "@pulumi/pulumi";
+import { input as inputs, output as outputs } from "./types";
+import * as utilities from "./utilities";
+
+import * as random from "@pulumi/random";
+
+export class Foo extends pulumi.CustomResource {
+    /**
+     * Get an existing Foo resource's state with the given name, ID, and optional extra
+     * properties used to qualify the lookup.
+     *
+     * @param name The _unique_ name of the resulting resource.
+     * @param id The _unique_ provider ID of the resource to lookup.
+     * @param opts Optional settings to control the behavior of the CustomResource.
+     */
+    public static get(name: string, id: pulumi.Input<pulumi.ID>, opts?: pulumi.CustomResourceOptions): Foo {
+        return new Foo(name, undefined as any, { ...opts, id: id });
+    }
+
+    /** @internal */
+    public static readonly __pulumiType = 'example::Foo';
+
+    /**
+     * Returns true if the given object is an instance of Foo.  This is designed to work even
+     * when multiple copies of the Pulumi SDK have been loaded into the same process.
+     */
+    public static isInstance(obj: any): obj is Foo {
+        if (obj === undefined || obj === null) {
+            return false;
+        }
+        return obj['__pulumiType'] === Foo.__pulumiType;
+    }
+
+
+    /**
+     * Create a Foo resource with the given unique name, arguments, and options.
+     *
+     * @param name The _unique_ name of the resource.
+     * @param args The arguments to use to populate this resource's properties.
+     * @param opts A bag of options that control this resource's behavior.
+     */
+    constructor(name: string, args?: FooArgs, opts?: pulumi.CustomResourceOptions) {
+        let inputs: pulumi.Inputs = {};
+        opts = opts || {};
+        if (!opts.id) {
+        } else {
+        }
+        if (!opts.version) {
+            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion()});
+        }
+        super(Foo.__pulumiType, name, inputs, opts);
+    }
+
+    /**
+     * A description of bar.
+     */
+    bar(args: Foo.BarArgs): pulumi.Output<Foo.BarResult> {
+        return pulumi.runtime.call("example::Foo/bar", {
+            "__self__": this,
```

Comment:
Nit: if there's never any renaming here, could we just pack this up into `withSelf(this, args)` or even `{'self': this, 'args': args}` and have `call` unpack. It's no big deal I guess since methods are relatively rare, I've just been cognizant of generated code size.